core/src/main/scala/kafka/server/DelayedOperation.scala

```diff
@@ -228,29 +211,37 @@ final class DelayedOperationPurgatory[T <: DelayedOperation](purgatoryName: Stri
     // if the operation is completed (by another thread) between the two tryComplete() calls, the
     // operation is unnecessarily added for watch. However, this is a less severe issue since the
     // expire reaper will clean it up periodically.
-
-    // At this point the only thread that can attempt this operation is this current thread
-    // Hence it is safe to tryComplete() without a lock
-    var isCompletedByMe = operation.tryComplete()
-    if (isCompletedByMe)
-      return true
-
-    var watchCreated = false
-    for(key <- watchKeys) {
-      // If the operation is already completed, stop adding it to the rest of the watcher list.
-      if (operation.isCompleted)
-        return false
-      watchForOperation(key, operation)
-
-      if (!watchCreated) {
-        watchCreated = true
-        estimatedTotalOperations.incrementAndGet()
+    //
+    // ==============[story about lock]==============
+    // There is a potential deadlock in practice if we don't hold the lock while adding the operation to watch
+    // list and do the final tryComplete() check. For example,
+    // 1) thread_a holds readlock of stateLock from TransactionStateManager
+    // 2) thread_a is executing tryCompleteElseWatch
+    // 3) thread_a adds op to watch list
+    // 4) thread_b requires writelock of stateLock from TransactionStateManager (blocked by thread_a)
+    // 5) thread_c holds lock of op (from watch list)
```

Comment:
thread_c holds lock of op => thread_c calls checkAndComplete () and holds lock of op